### PR TITLE
fix(expo,cli): export `@expo/metro-config` from controlling `@expo/cli` instead of using implicit dependency

### PR DESCRIPTION
**Pull request by byCedric**
# Why

With our change of import when customizing the Metro config (`@expo/metro-config` to `expo/metro-config`) we overlooked the implicit dependency issue. Basically what's happening is this:

- `expo` exports `@expo/metro-config` as `expo/metro-config`
- `expo` **does not** have `@expo/metro-config` as `dependencies` or `peerDependencies` and relies on `@expo/cli`'s version

For [isolated modules](https://github.com/npm/rfcs/blob/main/accepted/0042-isolated-mode.md), pnpm and future npm versions, this is a nightmare and won't work properly. `@expo/metro-config` is not defined as dependency to `expo`, and thus never linked in the eventual **node_modules** structure.

But even with Yarn (v1) there are cases where this does not work as we want it to. When users add a different version of `@expo/metro-config` to their project dependencies, the version `expo/metro-config` exports is not implicitly inherited from `@expo/cli`, but from the project dependencies.

# Yarn v1's case

The reason why the project dependencies are taking priority over the `@expo/cli`'s metro config version is the implicitness of the `@expo/metro-config` relation to `expo`.

When users don't have `@expo/metro-config` listed, the dependency chain `expo` -> `@expo/cli` -> `@expo/metro-config` gets hoisted to the root of the **node_modules** folder:

<details><summary>See <b>node_modules</b> structure information without <code>@expo/metro-config</code> project dependency</summary>

image

</details>

But, when users add a different version of `@expo/metro-config` as project dependency, this will be placed inside the root **node_modules** instead. While the correct metro config version is installed in the **node_modules/@expo/cli/node_modules/@expo/metro-config** folder.

<details><summary>See <b>node_modules</b> structure information with incorrect <code>@expo/metro-config</code> project dependency</summary>

image

</details>

# How

- Since `@expo/cli` should be "in control" of the compatible `@expo/metro-config`, I added the `@expo/cli/metro-config` export.
- Updated `expo/metro-config` to re-export `@expo/cli/metro-config` instead of `@expo/metro-config`

# Test Plan

- `$ yarn create expo ./test-metro-config -t tabs@49`
- `$ cd ./test-metro-config`
- `$ yarn add @expo/metro-config@0.0.1`
- `$ yarn expo start` -> should work fine with this change

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by byCedric** [comment]
TL;DR; implicit dependencies bad, need to have explicit dependencies for everything (probably even `expo/modules-core` too, where all `expo-*` packages add a peer dependency to `"expo": "*"`)

File: packages/@expo/cli/metro-config.js [hunk 1]
```diff
@@ -0,0 +1 @@
+module.exports = require('@expo/metro-config');
```
Comment (EvanBacon) [inline packages/@expo/cli/metro-config.js:1]: missing the `metro-config.d.ts`

**Comment by ide** [comment]
At a glance the plan in the "How" section LGTM

File: packages/@expo/cli/src/start/server/metro/resolveFromProject.ts [hunk 2]
```diff
@@ -44,9 +44,10 @@ export function importMetroHmrServerFromProject(
 }
 
 export function importExpoMetroConfig(projectRoot: string) {
-  return importFromProjectOrFallback<typeof import('@expo/metro-config')>(
+  // `@expo/metro-config` version is controlled by `@expo/cli` to avoid conflicting `@expo/metro-config` versions in project dependencies
+  return importFromProjectOrFallback<typeof import('@expo/cli/metro-config')>(
     projectRoot,
-    '@expo/metro-config'
+    '@expo/cli/metro-config'
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/metro/resolveFromProject.ts:50]: This doesn’t seem like it needs to change

File: packages/expo/metro-config.js [hunk 3]
```diff
@@ -1 +1 @@
-module.exports = require('@expo/metro-config');
\ No newline at end of file
+module.exports = require('@expo/cli/metro-config');
```
Comment (EvanBacon) [inline packages/expo/metro-config.js:1]: Could we not just add a dependency to the expo package? I’m not especially in love with allowing imports into expo/cli:
1. This makes it harder to ncc in the future.
2. Allowing imports to expo cli is how we ended up with xdl.

File: packages/@expo/cli/src/start/server/metro/resolveFromProject.ts (hunk 2 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/resolveFromProject.ts:50]: I thought it didn't either, but the reason why these also need to change it because they resolve from `projectRoot`. Basically, we treat all modules here as being a "list" inside `<project>/node_modules`, while technically we should treat it as a proper dependency graph.

In this case, what happens is `@expo/metro-config` is resolved from project root, or in other words `<project>/node_modules>`. Meaning that if the user, or any other library, added the incorrect `@expo/metro-config@0.0.1` version, we are still pulling in the wrong one.

The correct way of resolving this, and other libraries in this file, like how we would traverse dependency graphs. E.g.
- resolve `expo`
- use `expo` location to resolve `@expo/cli`
- use `@expo/cli` location to resolve `@expo/metro-config`

File: packages/expo/metro-config.js (hunk 3 above)
Comment (byCedric) [reply packages/expo/metro-config.js:1]: Yeah, we can also decide to move "ownership" to `expo` itself. Where ownership means "`expo` version dictates what versions of `@expo/metro-config` should be used for this SDK version".

**Comment by byCedric** [comment]
Superseded by https://github.com/expo/expo/pull/25804

**Comment by byCedric** [comment]
Fully superseded by #25804
